File: app/src/main/java/io/github/hrilab/pepper_realtime/YouTubeSearchService.java
package io.github.hrilab.pepper_realtime;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;

import androidx.annotation.NonNull;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class YouTubeSearchService {
    private static final String TAG = "YouTubeSearchService";
    private static final String YOUTUBE_API_BASE_URL = "https://www.googleapis.com/youtube/v3/search";
    
    private final String apiKey;
    private final OkHttpClient httpClient;
    
    public static class YouTubeVideo {
        private final String videoId;
        private final String title;
        private final String channelTitle;
        private final String thumbnailUrl;
        
        public YouTubeVideo(String videoId, String title, String channelTitle, String thumbnailUrl) {
            this.videoId = videoId;
            this.title = title;
            this.channelTitle = channelTitle;
            this.thumbnailUrl = thumbnailUrl;
        }
        
        public String getVideoId() { return videoId; }
        public String getTitle() { return title; }
        public String getChannelTitle() { return channelTitle; }
        @SuppressWarnings("unused")
        public String getThumbnailUrl() { return thumbnailUrl; }
        
        @NonNull
        @Override
        public String toString() {
            return "YouTubeVideo{" +
                    "videoId='" + videoId + '\'' +
                    ", title='" + title + '\'' +
                    ", channelTitle='" + channelTitle + '\'' +
                    '}';
        }
    }
    
    public YouTubeSearchService(String apiKey) {
        this.apiKey = apiKey;
        this.httpClient = OptimizedHttpClientManager.getInstance().getQuickApiClient();
    }
    
    /**
     * Search for YouTube videos based on query
     * @param query Search query (e.g. "like a virgin madonna")
     * @param maxResults Maximum number of results (default: 1 for first result)
     * @return First video result or null if no results
     */
    public YouTubeVideo searchVideo(String query, int maxResults) throws Exception {
        if (query == null || query.trim().isEmpty()) {
            throw new IllegalArgumentException("Search query cannot be empty");
        }
        
        if (apiKey == null || apiKey.isEmpty()) {
            throw new IllegalArgumentException("YouTube API key is required");
        }
        
        Log.i(TAG, "Searching YouTube for: " + query);
        
        // Build API URL
        String url = YOUTUBE_API_BASE_URL + 
                "?part=snippet" +
                "&type=video" +
                "&maxResults=" + maxResults +
                "&q=" + java.net.URLEncoder.encode(query, "UTF-8") +
                "&key=" + apiKey;
        
        // Make API request
        Request request = new Request.Builder()
                .url(url)
                .addHeader("Accept", "application/json")
                .build();
        
        try (Response response = httpClient.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                String errorBody = (response.body() != null) ? safeBodyString(response) : "Unknown error";
                Log.e(TAG, "YouTube API error: " + response.code() + " - " + errorBody);
                throw new IOException("YouTube API request failed: " + response.code());
            }
            
            String responseBody = (response.body() != null) ? safeBodyString(response) : null;
            Log.d(TAG, "YouTube API response received");
            
            // Parse JSON response
            if (responseBody == null) {
                Log.e(TAG, "Empty response body from YouTube API");
                return null;
            }
            JSONObject jsonResponse = new JSONObject(responseBody);
            JSONArray items = jsonResponse.getJSONArray("items");
            
            if (items.length() == 0) {
                Log.w(TAG, "No videos found for query: " + query);
                return null;
            }
            
            // Extract first video
            JSONObject firstItem = items.optJSONObject(0);
            if (firstItem == null) {
                Log.w(TAG, "Empty first item in YouTube response");
                return null;
            }
            JSONObject idObj = firstItem.optJSONObject("id");
            String videoId = idObj != null ? idObj.optString("videoId", "") : "";
            
            JSONObject snippet = firstItem.getJSONObject("snippet");
            String title = snippet.optString("title", "");
            String channelTitle = snippet.optString("channelTitle", "");
            
            // Get thumbnail URL (null-safe locals to satisfy inspections)
            JSONObject thumbnails = snippet.optJSONObject("thumbnails");
            JSONObject defaultThumb = thumbnails != null ? thumbnails.optJSONObject("default") : null;
            String thumbnailUrl = defaultThumb != null ? defaultThumb.optString("url", "") : "";
            
            YouTubeVideo video = new YouTubeVideo(videoId, title, channelTitle, thumbnailUrl);
            Log.i(TAG, "Found video: " + video);
            
            return video;
            
        } catch (Exception e) {
            Log.e(TAG, "Error searching YouTube", e);
            throw e;
        }
    }

    private static String safeBodyString(Response response) throws IOException {
        ResponseBody body = response.body();
        return body != null ? body.string() : null;
    }
    
    /**
     * Search for first video result
     * @param query Search query
     * @return First video result or null if no results
     */
    public YouTubeVideo searchFirstVideo(String query) throws Exception {
        return searchVideo(query, 1);
    }
    
    
    // Convenience overload removed as unused
}

File: app/src/main/java/io/github/anonymous/pepper_realtime/service/YouTubeSearchService.kt
package io.github.anonymous.pepper_realtime.service

import android.util.Log
import io.github.anonymous.pepper_realtime.di.IoDispatcher
import io.github.anonymous.pepper_realtime.network.HttpClientManager
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Request
import org.json.JSONObject
import java.io.IOException
import java.net.URLEncoder
import javax.inject.Inject

/**
 * Service for searching YouTube videos via the YouTube Data API v3.
 */
class YouTubeSearchService(
    private val apiKey: String
) {
    // Secondary constructor for DI (if needed in future)
    @Inject
    constructor(
        apiKey: String,
        @Suppress("UNUSED_PARAMETER") httpClientManager: HttpClientManager,
        @Suppress("UNUSED_PARAMETER") @IoDispatcher ioDispatcher: CoroutineDispatcher
    ) : this(apiKey)
    companion object {
        private const val TAG = "YouTubeSearchService"
        private const val YOUTUBE_API_BASE_URL = "https://www.googleapis.com/youtube/v3/search"
    }

    /**
     * Represents a YouTube video search result.
     */
    data class YouTubeVideo(
        val videoId: String,
        val title: String,
        val channelTitle: String,
        val thumbnailUrl: String
    ) {
        override fun toString(): String {
            return "YouTubeVideo{videoId='$videoId', title='$title', channelTitle='$channelTitle'}"
        }
    }

    private val httpClient by lazy { HttpClientManager.getInstance().getQuickApiClient() }

    /**
     * Search for YouTube videos based on query.
     * This is a suspend function - preferred for new Kotlin code.
     *
     * @param query Search query (e.g. "like a virgin madonna")
     * @param maxResults Maximum number of results (default: 1 for first result)
     * @return First video result or null if no results
     */
    suspend fun searchVideo(query: String, maxResults: Int = 1): YouTubeVideo? = withContext(Dispatchers.IO) {
        require(query.isNotBlank()) { "Search query cannot be empty" }
        require(apiKey.isNotEmpty()) { "YouTube API key is required" }

        Log.i(TAG, "Searching YouTube for: $query")

        // Build API URL
        val url = "$YOUTUBE_API_BASE_URL" +
                "?part=snippet" +
                "&type=video" +
                "&maxResults=$maxResults" +
                "&q=${URLEncoder.encode(query, "UTF-8")}" +
                "&key=$apiKey"

        val request = Request.Builder()
            .url(url)
            .addHeader("Accept", "application/json")
            .build()

        try {
            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    val errorBody = response.body?.string() ?: "Unknown error"
                    Log.e(TAG, "YouTube API error: ${response.code} - $errorBody")
                    throw IOException("YouTube API request failed: ${response.code}")
                }

                val responseBody = response.body?.string()
                Log.d(TAG, "YouTube API response received")

                if (responseBody == null) {
                    Log.e(TAG, "Empty response body from YouTube API")
                    return@withContext null
                }

                val jsonResponse = JSONObject(responseBody)
                val items = jsonResponse.getJSONArray("items")

                if (items.length() == 0) {
                    Log.w(TAG, "No videos found for query: $query")
                    return@withContext null
                }

                // Extract first video
                val firstItem = items.optJSONObject(0)
                if (firstItem == null) {
                    Log.w(TAG, "Empty first item in YouTube response")
                    return@withContext null
                }

                val idObj = firstItem.optJSONObject("id")
                val videoId = idObj?.optString("videoId", "") ?: ""

                val snippet = firstItem.getJSONObject("snippet")
                val title = snippet.optString("title", "")
                val channelTitle = snippet.optString("channelTitle", "")

                // Get thumbnail URL (null-safe)
                val thumbnails = snippet.optJSONObject("thumbnails")
                val defaultThumb = thumbnails?.optJSONObject("default")
                val thumbnailUrl = defaultThumb?.optString("url", "") ?: ""

                val video = YouTubeVideo(videoId, title, channelTitle, thumbnailUrl)
                Log.i(TAG, "Found video: $video")

                video
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error searching YouTube", e)
            throw e
        }
    }

    /**
     * Search for first video result.
     *
     * @param query Search query
     * @return First video result or null if no results
     */
    suspend fun searchFirstVideo(query: String): YouTubeVideo? = searchVideo(query, 1)
}


